model/src/util/image_hash.rs

```diff
@@ -0,0 +1,768 @@
+    /// efficient image hash.
+    ///
+    /// # Examples
+    ///
+    /// Parse an image hash and then check the packed bytes:
+    ///
+    /// ```
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// use twilight_model::util::ImageHash;
+    ///
+    /// let input = b"f49d812ca33c1cbbeec96b9f64487c7c";
+    /// let hash = ImageHash::parse(input)?;
+    /// let bytes = hash.bytes();
+    ///
+    /// // Byte correlates to 15 (`b'f'`) followed by 4 (`b'4'`), in total `f4`.
+    /// assert_eq!(0b1111_0100, bytes[0]);
+    ///
+    /// // Byte correlates to 7 (`b'7'`) followed by 12 (`b'c'`), in total `7c`.
+    /// assert_eq!(0b0111_1100, bytes[15]);
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// [`is_animated`]: Self::is_animated
+    /// [`new`]: Self::new
+    pub const fn bytes(self) -> [u8; 16] {
+        self.bytes
+    }
+
+    /// Whether the hash is for an animated image.
+    ///
+    /// # Examples
+    ///
+    /// Parse an animated image hash prefixed with `a_` and another static image
+    /// hash that is not prefixed with `a_`:
+    ///
+    /// ```
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// use twilight_model::util::ImageHash;
+    ///
+    /// let animated_input = "a_5145104ad8e8c9e765883813e4abbcc8";
+    /// let animated_hash = ImageHash::parse(animated_input.as_bytes())?;
+    /// assert!(animated_hash.is_animated());
+    ///
+    /// let static_input = "c7e7c4b8469d790cb9b293759e60953d";
+    /// let static_hash = ImageHash::parse(static_input.as_bytes())?;
+    /// assert!(!static_hash.is_animated());
+    /// # Ok(()) }
+    /// ```
+    pub const fn is_animated(self) -> bool {
+        self.animated
+    }
+
+    /// Create an iterator over the [nibbles] of the hexadecimal image hash.
+    ///
+    /// # Examples
+    ///
+    /// Parse an image hash and then iterate over the nibbles:
+    ///
+    /// ```
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// use twilight_model::util::ImageHash;
+    ///
+    /// let input = b"1d9811c4cd3782148915c522b02878fc";
+    /// let hash = ImageHash::parse(input)?;
+    /// let mut nibbles = hash.nibbles();
+    ///
+    /// assert_eq!(Some(b'1'), nibbles.next());
+    /// assert_eq!(Some(b'd'), nibbles.next());
+    /// assert_eq!(Some(b'c'), nibbles.nth(29));
+    /// assert!(nibbles.next().is_none());
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// [nibbles]: https://en.wikipedia.org/wiki/Nibble
+    pub const fn nibbles(self) -> Nibbles {
+        Nibbles::new(self)
+    }
+
+    /// Determine whether a haystack starts with a needle.
+    const fn starts_with(haystack: &[u8], needle: &[u8]) -> bool {
+        if needle.len() > haystack.len() {
+            return false;
+        }
+
+        let mut idx = 0;
+
+        while idx < needle.len() {
+            if haystack[idx] != needle[idx] {
+                return false;
+            }
+
+            idx += 1;
+        }
+
+        true
+    }
```

Comment:
This could just be `&haystack[..needle.len()] == needle`

Reply:
std::ops::Index isn't const